### PR TITLE
Adds 8-bit Integer Quantization Documentation

### PR DESCRIPTION
**Pull request by JyotinderSingh**

### CODE REVIEW

File: guides/int8_quantization_in_keras.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+# Compute a simple MSE between FP and INT8 outputs.
+mse = keras.ops.mean(keras.ops.square(y_fp32 - y_int8))
+print("Full-Precision vs INT8 MSE:", float(mse))
+
+
+"""
+It is evident that the INT8 quantized model produces outputs close to the original FP32
+model, as indicated by the low MSE value.
+
+## Saving and reloading a quantized model.
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done

File: guides/int8_quantization_in_keras.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+# Save the quantized model and reload to verify round-trip.
+model.save("int8.keras")
+int8_reloaded = saving.load_model("int8.keras")
+y_int8_reloaded = int8_reloaded(x_eval)
+roundtrip_mse = keras.ops.mean(keras.ops.square(y_int8 - y_int8_reloaded))
+print("MSE (INT8 vs reloaded-INT8):", float(roundtrip_mse))
+
+"""
+## Quantizing a KerasHub model.
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done.

File: guides/int8_quantization_in_keras.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+All KerasHub models support the `.quantize(...)` API for post-training quantization,
+and follow the same workflow as above.
+
+In this example, we will:
+
+1. Load the [gemma3_1b](https://www.kaggle.com/models/keras/gemma3/keras/gemma3_1b)
+preset from KerasHub
+1. Generate text using both the full-precision and quantized models, and compare outputs.
+1. Save both models to disk and compute storage savings.
+1. Reload the INT8 model and verify output consistency with the original quantized model.
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Fixed numbering, did not remove blank line since markdown conventions say lists should be surrounded by blank lines

File: guides/int8_quantization_in_keras.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+* Integer domain: `[-128, 127]` (256 levels).
+* For a tensor (often per-output-channel for weights) with values `w`:
+  * Compute `a_max = max(abs(w))`.
+  * Set scale `s = (2 * a_max) / 256`.
+  * Quantize: `q = clip(round(w / s), -128, 127)` (stored as int8) and keep `s`.
+* Inference uses `q` and `s` to reconstruct effective weights on the fly
+(`w ≈ s · q`) or folds `s` into the matmul/conv for efficiency.
+
+### Trade-off
+Wider dynamic range (larger `a_max`) reduces clipping but increases rounding error;
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: nitpick: blank line in-between

File: guides/int8_quantization_in_keras.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+### How it works
+
+Quantization maps real values to 8-bit integers with a scale:
+
+* Integer domain: `[-128, 127]` (256 levels).
+* For a tensor (often per-output-channel for weights) with values `w`:
+  * Compute `a_max = max(abs(w))`.
+  * Set scale `s = (2 * a_max) / 256`.
+  * Quantize: `q = clip(round(w / s), -128, 127)` (stored as int8) and keep `s`.
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: If scale really is `(2 * a_max) / 256` there is no clipping.

Later you talk about wider range / tighter range and the effect on clipping. So should you introduce the notion that there is a factor between `(2 * a_max) / 256` and the scale you actually pick?

File: guides/int8_quantization_in_keras.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+* Inference uses `q` and `s` to reconstruct effective weights on the fly
+(`w ≈ s · q`) or folds `s` into the matmul/conv for efficiency.
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: indent

File: guides/int8_quantization_in_keras.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+* Memory / bandwidth bound models: When implementation spends most of its time on memory I/O,
+reducing the computation time does not reduce their overall runtime. `int8` reduces bytes
+moved by ~4x vs `float32`, improving cache behavior and reducing memory stalls;
+this often helps more than increasing raw FLOPs.
+* Compute bound layers on supported hardware: On NVIDIA GPUs, int8
+[Tensor Cores](https://www.nvidia.com/en-us/data-center/tensor-cores/) speed up matmul/conv,
+boosting throughput on compute-limited layers.
+* Accuracy: Many models retain near-FP accuracy with `float16`; `int8` may introduce a modest
+drop (often ~1-5% depending on task/model/data). Always validate on your own dataset.
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: indent lines that don't have a * so that they line up:
```
* Memory ...
  reducing ...
```

File: guides/int8_quantization_in_keras.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+### What Keras does in `int8` mode
+
+* **Mapping**: Symmetric, linear quantization with `int8` plus a floating-point scale.
+* **Weights**: per-output-channel scales to preserve accuracy.
+* **Activations**: **dynamic AbsMax** scaling computed at runtime.
+* **Graph rewrite**: Quantization is applied after weights are trained and built; the graph
+is rewritten so you can run or save immediately.
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: indent

File: guides/int8_quantization_in_keras.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+Accelerator: GPU
+"""
+
+"""
+## What is INT8 quantization?
+
+Quantization lowers the numerical precision of weights and activations to reduce memory use
+and often speed up inference, at the cost of a small accuracy drop. Moving from `float32` to
+`float16` halves the memory requirements; `float32` to `int8` is ~4x smaller (and ~2x vs
+`float16`). On hardware with low-precision kernels (e.g., Tensor Cores), this can also
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: Which Tensor Cores are you referring to? The NVidia ones? TPUs have TensorCores too, so you should disambiguate.

File: guides/int8_quantization_in_keras.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
+"""
+Title: 8-bit Integer Quantization in Keras
+Author: [Jyotinder Singh](https://x.com/Jyotinder_Singh)
+Date created: 2025/10/14
+Last modified: 2025/10/14
+Description: Complete guide to using INT8 quantization in Keras and KerasHub
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: In some places it's INT8, other places it's int8, and some other places `int8` with backquotes. Let's standardize.

File: guides/int8_quantization_in_keras.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+and then compare outputs with an MSE metric.
+"""
+
+import os
+import numpy as np
+import keras
+from keras import layers
+
+
+np.random.seed(7)
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: Any reason to seed like that?

File: guides/int8_quantization_in_keras.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+It is evident that the INT8 quantized model produces outputs close to the original FP32
+model, as indicated by the low MSE value.
+
+## Saving and reloading a quantized model
+
+You can use the standard Keras saving and loading APIs with quantized models. Quantization
+is preserved when saving to `.keras` and loading back.
+"""
+
+from keras import saving
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: nitpick: it's only used once, so just do keras.saving.load_model(...)

File: guides/int8_quantization_in_keras.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+# Build a functional model.
+inputs = keras.Input(shape=(10,))
+x = layers.Dense(32, activation="relu")(inputs)
+outputs = layers.Dense(1, name="target")(x)
+model = keras.Model(inputs, outputs)
+model.build((None, 10))
+
+# Quantize the model in-place to INT8.
+model.quantize("int8")
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: Why not continue with the model you already have and that's already quantized?

File: guides/int8_quantization_in_keras.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+"""
+## Quantizing a KerasHub model
+
+All KerasHub models support the `.quantize(...)` API for post-training quantization,
+and follow the same workflow as above.
+
+In this example, we will:
+
+1. Load the [gemma3_1b](https://www.kaggle.com/models/keras/gemma3/keras/gemma3_1b)
+preset from KerasHub
```
Comment (hertschuh) [inline guides/int8_quantization_in_keras.py]: indent

File: guides/md/int8_quantization_in_keras.md [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+print(f"Gemma3: FP32 file size: {bytes_to_mib(gemma_fp32_size):.2f} MiB")
+print(f"Gemma3: INT8 file size: {bytes_to_mib(gemma_int8_size):.2f} MiB")
+print(f"Gemma3: Size reduction: {gemma_reduction:.1f}%")
+```
+
+<div class="k-default-codeblock">
+```
+Full-precision output: Keras is a deep learning library for Python. It is a high-level API for neural networks. It is a Python library for deep learning
+
+Quantized output: Keras is a deep learning library for Python. It is a high-level API for neural networks. It is a Python library for deep learning
```
Comment (hertschuh) [inline guides/md/int8_quantization_in_keras.md]: Hmm... is it actually really working :)  How far do you have to go to see a difference?

File: guides/int8_quantization_in_keras.py (hunk 4 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done

File: guides/int8_quantization_in_keras.py (hunk 10 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done.

File: guides/int8_quantization_in_keras.py (hunk 9 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done.

File: guides/int8_quantization_in_keras.py (hunk 6 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done.

File: guides/int8_quantization_in_keras.py (hunk 7 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done.

File: guides/int8_quantization_in_keras.py (hunk 8 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done.

File: guides/int8_quantization_in_keras.py (hunk 12 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done.

File: guides/int8_quantization_in_keras.py (hunk 14 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: Done.

File: guides/int8_quantization_in_keras.py (hunk 13 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: I was trying to keep each snippet self contained, but I guess that's sort of repetitive. Reusing the existing model now.

File: guides/md/int8_quantization_in_keras.md (hunk 15 above)
Comment (JyotinderSingh) [reply guides/md/int8_quantization_in_keras.md]: Great catch! I looked a little deeper and turns out the graph wasn't being reset, which is why it was giving different outputs on reload. I've opened a fix for this in the following PR: https://github.com/keras-team/keras-hub/pull/2436

I installed that change locally and regenerated the outputs for these docs as well. The outputs are now a little different after quantization, as expected.

File: guides/int8_quantization_in_keras.py (hunk 11 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: no particular reason, removed the seed.

File: guides/int8_quantization_in_keras.py (hunk 5 above)
Comment (JyotinderSingh) [reply guides/int8_quantization_in_keras.py]: I added the trad-off section just for some additional theory. But I agree it introduces some ambiguity with respect to what's actually happening (always clipping with `a_max`).
Removed it.
